### PR TITLE
IGNITE-11926: GridJobProcessor metrics migration.

### PR DESCRIPTION
**Pull request by nizhikov**

### CODE REVIEW

File: modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+    private static CountDownLatch latch;
+
+    /** Test correct calculation of rejected and waiting metrics of the {@link GridJobProcessor}. */
+    @Test
+    public void testGridJobWaitingRejectedMetrics() throws Exception {
+        latch = new CountDownLatch(1);
+
+        GridTestCollision collisioinSpi = new GridTestCollision();
+
+        IgniteConfiguration cfg = new IgniteConfiguration()
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java]: Use `getConfiguration()` method, which uses shared VM IP finder. `new IgniteConfiguration()` can affect other nodes, located on other machines.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+
+            SimplestTask task1 = new SimplestTask();
+            SimplestTask task2 = new SimplestTask();
+            SimplestTask task3 = new SimplestTask();
+
+            task1.block = true;
+            task2.block = true;
+            task3.block = true;
+
+            //Task will become "waiting", because of CollisionSpi implementation.
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java]: Space after //

File: modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+            ComputeTaskFuture<?> fut3 = g.compute().executeAsync(task3, 1);
+
+            assertEquals(0, started.value());
+            assertEquals(0, active.value());
+            assertEquals(3, waiting.value());
+            assertEquals(0, canceled.value());
+            assertEquals(0, rejected.value());
+            assertEquals(0, finished.value());
+
+            //Activating 2 of 3 jobs. Rejecting 1 of them.
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java]: Space after //

File: modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+            Iterator<CollisionJobContext> iter = collisioinSpi.jobs.values().iterator();
+
+            iter.next().cancel();
+
+            assertEquals(1, rejected.value());
+
+            Thread.sleep(100); //Sleeping to make sure totalWaitingTime will become more the zero.
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java]: Space after //

File: modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+
+            iter.next().activate();
+            iter.next().activate();
+
+            boolean res = waitForCondition(() -> active.value() > 0, TIMEOUT);
+
+            assertTrue(res);
+
+            Thread.sleep(100); //Sleeping to make sure totalExecutionTime will become more the zero.
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java]: Space after //

File: modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+            res = waitForCondition(() -> fut1.isDone() && fut2.isDone() && fut3.isDone(), TIMEOUT);
+
+            assertTrue(res);
+
+            res = waitForCondition(() -> finished.value() == 3, TIMEOUT);
+
+            assertTrue(res);
+
+            assertTrue("Execution time should be greater then zero.", totalExecutionTime.value() > 0);
+            assertTrue("Waiting time should be greater then zero.", totalWaitingTime.value() > 0);
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java]: It's better to check `totalWaitingTime` earlier, when `totalExecutionTime` is zero

File: modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+            assertEquals(0, rejected.value());
+            assertEquals(0, finished.value());
+            assertEquals(0, totalExecutionTime.value());
+            assertEquals(0, totalWaitingTime.value());
+
+            SimplestTask task = new SimplestTask();
+
+            g.compute().execute(task, 1);
+
+            //Waiting task to finish.
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java]: Space after //

File: modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+            boolean res = waitForCondition(() -> active.value() == 0, TIMEOUT);
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java]: Is there a race possible? `active == 0` can be in two cases: when the job is not activated yet and after the job is finished. Here you wait for `active == 0` and then checks for `finished == 1`. Does job always activated after `execute()` in this test?

File: modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java [hunk 9]
```diff
@@ -0,0 +1,358 @@
...
+
+            assertTrue("Active = " + active.value(), res);
+
+            assertEquals(1, started.value());
+            assertEquals(0, waiting.value());
+            assertEquals(0, canceled.value());
+            assertEquals(0, rejected.value());
+            assertEquals(1, finished.value());
+
+            //Task should block until latch is down.
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java]: Space after //

File: modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+            task.block = true;
+
+            ComputeTaskFuture<?> fut = g.compute().executeAsync(task, 1);
+
+            //Waiting task to start execution.
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java]: Space after //

File: modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java [hunk 11]
```diff
@@ -0,0 +1,358 @@
...
+
+            assertTrue("Active = " + active.value(), res);
+
+            assertEquals(2, started.value());
+            assertEquals(0, waiting.value());
+            assertEquals(0, canceled.value());
+            assertEquals(0, rejected.value());
+            assertEquals(1, finished.value());
+
+            res = waitForCondition(() -> active.value() > 0, TIMEOUT);
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java]: There nothing happened between wait for `active.value() == 1` and wait for `active.value() > 0`. Why should we wait twice?

File: modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java [hunk 12]
```diff
@@ -0,0 +1,358 @@
...
+            assertTrue(res);
+
+            Thread.sleep(100); //Sleeping to make sure totalExecutionTime will become more the zero.
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java]: Space after //

File: modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java [hunk 13]
```diff
@@ -0,0 +1,358 @@
...
+
+            //After latch is down, task should finish.
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java]: Space after //

File: modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java [hunk 14]
```diff
@@ -0,0 +1,358 @@
...
+
+            assertTrue("Active = " + active.value(), res);
+
+            assertEquals(3, started.value());
+            assertEquals(0, waiting.value());
+            assertEquals(0, canceled.value());
+            assertEquals(0, rejected.value());
+            assertEquals(2, finished.value());
+
+            //First cancel task, then allow it to finish.
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java]: Space after //

File: modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java [hunk 15]
```diff
@@ -0,0 +1,358 @@
...
+            res = waitForCondition(() -> finished.value() == 3, TIMEOUT);
+
+            assertTrue("Finished = " + finished.value(), res);
+        }
+    }
+
+    /** */
+    private static class SimplestJob implements ComputeJob {
+        /** */
+        boolean block;
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java]: final

File: modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java [hunk 16]
```diff
@@ -0,0 +1,358 @@
...
+            }
+
+            return "1";
+        }
+    }
+
+    /** */
+    private static class SimplestTask extends ComputeTaskAdapter<Object, Object> {
+        /** */
+        boolean block;
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java]: volatile or final (better)

File: modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java [hunk 17]
```diff
@@ -0,0 +1,358 @@
...
+
+/**
+ * Grid job metrics processor load test.
+ */
+public class GridJobMetricsSelfTest extends GridCommonAbstractTest {
+    /** */
+    public static final long TIMEOUT = 10_000;
+
+    /** */
+    private static CountDownLatch latch;
```
Comment (alex-plekhanov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java]: volatile

File: modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java (hunk 6 above)
Comment (nizhikov) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java]: Check moved to the line 143

File: modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java (hunk 8 above)
Comment (nizhikov) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java]: In this test `CollisionSpi` is null.
So `GridJobProcessor#jobAlwaysActivate` will be true.
So yes, Ignite will always activate job on `g.compute().execute(task, 1);` invocation.
For more details, please, see `GridJobProcessor#processJobExecuteRequest` line 1268 in this PR and `GridJobProcessor#onBeforeActivateJob` method.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java (hunk 11 above)
Comment (nizhikov) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java]: 1. `ComputeTaskFuture<?> fut = g.compute().executeAsync(task, 1);` - starting task invocation that will be waiting on `latch` for imitaion of long running job.
2. `res = waitForCondition(() -> active.value() == 1, TIMEOUT);` - waiting this job to start.
3. `latch.countDown();` - now started job can unblock and finish.
4. `res = waitForCondition(() -> active.value() == 0, TIMEOUT);` - waiting for metric update.

Comment (nizhikov) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/jobmetrics/GridJobMetricsSelfTest.java]: OK, now I see - the question is about the second assert of `active.value() > 0` :)
Seems, it's duplicate, as you wrote.
Removed.
Thanks!

**Pull request merged**